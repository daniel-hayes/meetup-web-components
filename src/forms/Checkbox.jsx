import PropTypes from 'prop-types';
import React from 'react';
import { default as SwarmCheckbox } from '@meetup/swarm-components/lib/Checkbox';

/**
 * @module Checkbox
 */
class Checkbox extends React.PureComponent {
	constructor(props) {
		super(props);

		this.state = {
			checked: props.checked || false,
		};

		this.onChange = this.onChange.bind(this);
	}

	componentWillReceiveProps(nextProps) {
		const { checked, controlled } = nextProps;

		if (this.props.controlled !== controlled && controlled) {
			this.setState({ checked });
		}
	}

	onChange(e) {
		this.props.onChange && this.props.onChange(e);

		if (this.props.controlled) {
			this.setState({ checked: e.target.checked });
		}
	}

	getChecked() {
		if (this.props.controlled) {
			return this.state.checked;
		}

		return this.props.checked;
	}

	render() {
		const {
			checked,
			children,
			id,
			label,
			disabled,
			name, // TODO: should v2 Checkbox support 'name'?
			value, // TODO: Should v2 Checkbox support 'value'?
			// iconShape, // TODO: should v2 Checkbox support 'iconShape'?
			...other
		} = this.props;

		const elId = id || `${name}-${value}`;
		const stateChecked = this.getChecked();

		return children ? (
			<SwarmCheckbox
				checked={stateChecked}
				label={label}
				id={elId}
				disabled={disabled}
				onChange={this.onChange}
				{...other}
			>
				{children}
			</SwarmCheckbox>
		) : (
			<SwarmCheckbox
				checked={checked}
				label={label}
				id={elId}
				disabled={disabled}
				onChange={this.onChange}
				{...other}
			/>
		);
	}
}

Checkbox.propTypes = {
	checked: PropTypes.bool,
	disabled: PropTypes.bool,

	/** Whether the input's `checked` prop is handled outside the component or by using component's internal state */
	controlled: PropTypes.bool,

	/** Adds an `id` attribute to the input, and associates it with the `<label />` */
	id: PropTypes.string,

	/** What we render into the input's `<label />` */
	label: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),

	/** The `name` attribute for the input */
	name: PropTypes.string.isRequired,

	/** The `value` attribute for the input */
	value: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]), // checkboxes don't need values

	/** Callback that happens when the input changes */
	onChange: PropTypes.func,

	/** Optional icon for checked state, defaults to 'check'  */
	iconShape: PropTypes.string,
};

Checkbox.defaultProps = {
	controlled: true,
};

export default Checkbox;
