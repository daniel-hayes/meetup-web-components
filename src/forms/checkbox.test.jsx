import React from 'react';
import { shallow, render, mount } from 'enzyme';

import Checkbox, {
	FAUX_TOGGLE_CLASS,
	FOCUSED_CHECKBOX_CLASS,
	DISABLED_CHECKBOX_CLASS,
} from './Checkbox';

const fakeOnChange = jest.fn();

const JSXcheckboxUnchecked = (
	<Checkbox label="Hello!" name="greeting" id="hello" value="hello" />
);
const JSXcheckboxChecked = (
	<Checkbox label="Hello!" name="greeting" id="hello" value="hello" checked />
);
const JSXcheckboxDisabled = (
	<Checkbox label="Hello!" name="greeting" id="hello" value="hello" disabled />
);
const JSXcheckboxUncheckedWithOnChange = (
	<Checkbox
		label="Hello!"
		name="greeting"
		id="hello"
		value="hello"
		onChange={fakeOnChange}
	/>
);

describe('Checkbox', function() {
	describe('basic', function() {
		const uncheckedComponent = shallow(JSXcheckboxUnchecked);
		const checkedComponent = shallow(JSXcheckboxChecked);

		it('renders unchecked checkbox', function() {
			expect(uncheckedComponent).toMatchSnapshot();
		});

		it('renders checked checkbox', function() {
			expect(checkedComponent).toMatchSnapshot();
		});

		it('has a label with correct for attribute', () => {
			const component = render(JSXcheckboxUnchecked);
			expect(component.attr('for')).toEqual('hello');
		});

		it('has a correct for attribute with generated id if no id is given', function() {
			const component = render(JSXcheckboxUnchecked);
			const checkbox = component.find('input');

			expect(component.attr('for')).toEqual('hello');
			expect(checkbox.attr('id')).toEqual('hello');
		});

		it('calls onChange and sets state clicked', function() {
			fakeOnChange.mockClear();
			const component = mount(JSXcheckboxUncheckedWithOnChange);
			const checkbox = component.find('input');

			expect(checkbox.props().checked).toBe(false);
			expect(fakeOnChange).not.toHaveBeenCalled();
			checkbox.simulate('change', { target: { checked: true } });
			expect(component.find('input').props().checked).toBe(true);
			expect(component.instance().state.checked).toBe(true);
			expect(fakeOnChange).toHaveBeenCalled();
		});

		it('calls onChange and does not set state', function() {
			fakeOnChange.mockClear();
			const component = mount(
				<Checkbox
					label="Hello!"
					name="greeting"
					id="hello"
					value="hello"
					checked
					controlled={false}
					onChange={fakeOnChange}
				/>
			);
			const checkbox = component.find('input');

			expect(checkbox.props().checked).toBe(true);
			expect(fakeOnChange).not.toHaveBeenCalled();
			checkbox.simulate('change', { target: { checked: true } });
			expect(component.find('input').props().checked).toBe(true);
			expect(component.instance().state.checked).toBe(true);
			expect(fakeOnChange).toHaveBeenCalled();
		});

		it('sets state when changing from uncontrolled to controlled', function() {
			const component = mount(
				<Checkbox
					label="Hello!"
					name="greeting"
					id="hello"
					value="hello"
					checked={false}
					controlled={false}
				/>
			);
			expect(component.instance().state.checked).toBe(false);
			component.setProps({ checked: true });
			expect(component.instance().state.checked).toBe(false);
			component.setProps({ controlled: true });
			expect(component.instance().state.checked).toBe(true);
		});
	});

	describe('fauxInput', function() {
		it('should render an icon when the faux input is checked', () => {
			const component = mount(JSXcheckboxChecked);
			const icon = component.find('svg');

			expect(icon.length).toBe(1);
		});

		it(`should add class '${FOCUSED_CHECKBOX_CLASS}' when the faux input is focused`, () => {
			const component = mount(JSXcheckboxUnchecked);
			const fauxInput = component.find(`.${FAUX_TOGGLE_CLASS}`);
			const input = component.find('input');

			expect(fauxInput.hasClass(FOCUSED_CHECKBOX_CLASS)).toBe(false);
			input.simulate('focus');
			expect(
				component.find(`.${FAUX_TOGGLE_CLASS}`).hasClass(FOCUSED_CHECKBOX_CLASS)
			).toBe(true);
		});

		it(`should remove class '${FOCUSED_CHECKBOX_CLASS}' when the faux input loses focus`, () => {
			const component = mount(JSXcheckboxUnchecked);
			const input = component.find('input');

			input.simulate('focus');
			expect(
				component.find(`.${FAUX_TOGGLE_CLASS}`).hasClass(FOCUSED_CHECKBOX_CLASS)
			).toBe(true);
			input.simulate('blur');
			expect(
				component.find(`.${FAUX_TOGGLE_CLASS}`).hasClass(FOCUSED_CHECKBOX_CLASS)
			).toBe(false);
		});

		it(`should add class ${DISABLED_CHECKBOX_CLASS} when the faux input is disabled`, () => {
			const component = mount(JSXcheckboxDisabled);
			const fauxInput = component.find(`.${FAUX_TOGGLE_CLASS}`);

			expect(fauxInput.hasClass(DISABLED_CHECKBOX_CLASS)).toBe(true);
		});
	});
});
