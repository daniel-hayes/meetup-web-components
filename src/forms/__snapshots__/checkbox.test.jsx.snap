// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox basic renders checked checkbox 1`] = `
<label
  className="toggleLabel label--minor display--block"
  htmlFor="hello"
>
  <Flex
    align="center"
    className="minTouchHeight"
  >
    <FlexItem
      shrink={true}
    >
      <input
        checked={true}
        className="checkbox visibility--a11yHide"
        id="hello"
        name="greeting"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        type="checkbox"
        value="hello"
      />
      <span
        className="display--block align--center fauxToggle fauxToggle--checkbox checked"
      >
        <Icon
          className="display--flex flex--center checkbox-indicator"
          shape="check"
          size="xs"
        />
      </span>
    </FlexItem>
    <FlexItem
      className="toggleLabel-container"
    >
      <span
        className="text--bold"
      >
        Hello!
      </span>
    </FlexItem>
  </Flex>
</label>
`;

exports[`Checkbox basic renders unchecked checkbox 1`] = `
<label
  className="toggleLabel label--minor display--block"
  htmlFor="hello"
>
  <Flex
    align="center"
    className="minTouchHeight"
  >
    <FlexItem
      shrink={true}
    >
      <input
        checked={false}
        className="checkbox visibility--a11yHide"
        id="hello"
        name="greeting"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        type="checkbox"
        value="hello"
      />
      <span
        className="display--block align--center fauxToggle fauxToggle--checkbox"
      />
    </FlexItem>
    <FlexItem
      className="toggleLabel-container"
    >
      <span
        className=""
      >
        Hello!
      </span>
    </FlexItem>
  </Flex>
</label>
`;
